mii/grpc_related/modelresponse_server.py

```diff
@@ -191,40 +199,82 @@ def intercept_service(self, continuation, handler_call_details):
         next_handler = continuation(handler_call_details)
         assert next_handler.unary_unary is not None
 
+        #USE KWARGS LIKE THEY ARE USED TO MAKE SESSIONS TO GET THE DEPLOYMENT NAME TO HASH THE COUNTERS/STUBS
+
         def invoke_intercept_method(request_proto, context):
             method_name = _get_grpc_method_name(handler_call_details.method)
-
             if method_name == TERMINATE_METHOD:
-                for stub in self.stubs:
-                    stub.invoke(TERMINATE_METHOD,
-                                google_dot_protobuf_dot_empty__pb2.Empty())
+                for deployment in self.stubs:
+                    for stub in self.stubs[deployment]:
+                        stub.invoke(TERMINATE_METHOD,
+                                    google_dot_protobuf_dot_empty__pb2.Empty())
                 self.asyncio_loop.call_soon_threadsafe(self.asyncio_loop.stop)
                 return next_handler.unary_unary(request_proto, context)
+            kwargs = unpack_proto_query_kwargs(request_proto.query_kwargs)
+            if method_name != TERMINATE_METHOD:
+                assert "deployment_name" in kwargs, "Must include deployment_name in kwargs for query"
+            deployment_name = kwargs.get('deployment_name')
+            kwargs.pop('deployment_name', None)
+            task = None
+            for repl in self.replica_configs:
+                if repl.deployment_name == deployment_name:
+                    task = repl.task
+                    break
+            method = GRPC_METHOD_TABLE[get_task(task)]
+            new_request = None
+            if method_name == "ConversationalReply":
+                request_dict = {}
+                request_dict['text'] = str(request_proto.text)
+                val = getattr(request_proto, 'conversation_id')
+                request_dict['conversation_id'] = int(val) if val is not None else None
+                request_dict['past_user_inputs'] = list(request_proto.past_user_inputs)
+                request_dict['generated_responses'] = list(
+                    request_proto.generated_responses)
+                new_request = method.pack_request_to_proto(request_dict, **kwargs)
+
+            elif method_name == "QuestionAndAnswerReply":
+                request_dict = {}
+                request_dict['question'] = str(request_proto.question)
+                request_dict['context'] = str(request_proto.context)
+                new_request = method.pack_request_to_proto(request_dict, **kwargs)
+            else:
+                request_dict = {}
+                request_dict["query"] = list(
+                    request_proto.request
+                ) if method_name == "GeneratorReply" or method_name == "Txt2ImgReply" else str(
+                    request_proto.request)
+                new_request = method.pack_request_to_proto(request_dict, **kwargs)
```

Comment:
This code looks like it is copied from `methods_table.py`:
https://github.com/microsoft/DeepSpeed-MII/blob/ce80314c0d086485b1559663102c48d2c24cb2eb/mii/method_table.py#L218

Can we find a way to reuse the code blocks we already have that handle this work?